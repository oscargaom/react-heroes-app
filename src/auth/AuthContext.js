import { createContext } from "react";

/*  Al usar createContext este nos facilita información que puede ser 
    compartida con los demás componentes del árbol de componentes */
export const AuthContext = createContext();