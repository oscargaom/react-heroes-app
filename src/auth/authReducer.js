import { types } from "../types/types";

export const authReducer = (state = {}, action) => {
    
    // console.log('=========================')
    // console.log('authReducer');
    // console.log('state');
    // console.log(state);
    // console.log('action');
    // console.log(action);
    // console.log('=========================')
    
    switch (action?.type) {
        case types.login:
            return {
                ...action.payload, 
                logged: true
            }
        case types.logout:
            return {logged: false};
        default:
            return state;
    }
};
