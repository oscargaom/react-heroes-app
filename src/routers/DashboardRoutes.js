import React from 'react'
import { Navbar } from '../componentes/ui/Navbar'
import { Switch, Route, Redirect } from 'react-router-dom'
import { MarvelScreen } from '../componentes/marvel/MarvelScreen'
import { HeroeScreen } from '../componentes/heroes/HeroeScreen'
import { DcScreen } from '../componentes/dc/DcScreen'
import { SearchScreen } from '../componentes/search/SearchScreen'

// export const DashboardRoutes = ({ history }) => {
export const DashboardRoutes = (props) => {
    // console.log('=======================');
    // console.log('DashboardRoutes.history');
    // console.log({ history });
    // console.log(props);

    return (
        <>
            {/* Como Navbar es un componente que no se encuentra dentro de una rutas (Route) 
                este no tiene acceso a las props por eso hay que pasarleas */}
            {/* <Navbar history={history} /> */}
            <Navbar />

            <div className="container mt-3">
                <Switch>
                    <Route exact path="/marvel" component={MarvelScreen} />
                    <Route exact path="/hero/:heroId" component={HeroeScreen} />
                    <Route exact path="/dc" component={DcScreen} />
                    <Route exact path="/search" component={SearchScreen} />
                    <Redirect to="/marvel" />
                </Switch>
            </div>
        </>
    )
}
