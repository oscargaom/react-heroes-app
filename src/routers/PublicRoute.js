import React from 'react'
import PropTypes from 'prop-types'

import { Route, Redirect } from "react-router-dom";

export const PublicRoute = ({isAuthenticated, component: Component, ...rest}) => {

    // console.log('=====================');
    // console.log('isAuthenticated');
    // console.log(isAuthenticated);
    // console.log('component');
    // console.log(Component);
    // console.log('rest');
    // console.log(rest);

    return(
        <Route {...rest} 
            component={props => (
                (!isAuthenticated)
                ? <Component {...props} />
                : <Redirect to="/" />
            )}
        />
    );
};

PublicRoute.propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    component: PropTypes.func.isRequired
}