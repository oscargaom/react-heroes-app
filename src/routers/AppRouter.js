import React, { useContext } from 'react';

import {
    BrowserRouter as Router,
    Switch
} from "react-router-dom";

import { LoginScreen } from '../componentes/login/LoginScreen';
import { DashboardRoutes } from './DashboardRoutes';
import { PrivateRoute } from './PrivateRoute'
import { AuthContext } from '../auth/AuthContext';
import { PublicRoute } from './PublicRoute';


export const AppRouter = () => {

    const {authUser:{logged}} = useContext(AuthContext);

    return (
        <div>
            <Router>
                <div>
                    {/* A <Switch> looks through its children <Route>s and
                        renders the first one that matches the current URL. */}
                    <Switch>
                        
                        {/* <Route exact path="/login" component={LoginScreen} /> */}
                        <PublicRoute exact 
                                    path="/login" 
                                    component={LoginScreen} 
                                    isAuthenticated={logged}            
                        />

                        {/* <Route path="/" component={DashboardRoutes} /> */}
                        <PrivateRoute 
                            path="/" 
                            component={DashboardRoutes} 
                            isAuthenticated={logged} />
                    </Switch>
                </div>
            </Router>
        </div>
    )
}
