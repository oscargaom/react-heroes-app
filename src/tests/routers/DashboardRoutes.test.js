import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { DashboardRoutes } from '../../routers/DashboardRoutes';
import { AuthContext } from '../../auth/AuthContext';
import { MemoryRouter } from 'react-router-dom';
import { authContextNotLoggedValue } from '../fixtures/authContextValue';

describe('Pruebas en el componente <DashboardRoutes />', () => {
    test('debe mostrarse el componenete correctamente', () => {

        // const contextValue = {
        //     authUser: {
        //         name: "Manuel",
        //         logged: false
        //     },
        //     authDispatch: jest.fn(),
        // };

        /*  
        */

        /*  Nos aparece este error sino se coloca <MemoryRouter />: Invariant failed: You should not use 
            <Switch> outside a <Router>.
            Cuando se ejecuta esta prueba y pasa por el componente <DashboardRoutes /> se hace la
            evaluación en el componente <Switch /> y a su vez en cada <Route exact path=".." component={..} />
            el cual regresa <Redirect to="/marvel" /> porque el path es igual a vacío , por eso en la salida 
            se ve el componente de marvel, pero si fuera <Redirect to="/dc" /> mostraría el componente de dc y lo
            mismo aplicaría si el path fuera "/search" o "/hero/:heroId".
        */

        const wrapper = mount(
            <AuthContext.Provider value={
                authContextNotLoggedValue
            } >
                <MemoryRouter>
                    <DashboardRoutes />
                </MemoryRouter>
            </AuthContext.Provider>
        );

        expect(wrapper.html()).toMatchSnapshot();
        expect((wrapper.find('.text-info').text().trim()))
            .toBe(authContextNotLoggedValue.authUser.name);
    });
})

