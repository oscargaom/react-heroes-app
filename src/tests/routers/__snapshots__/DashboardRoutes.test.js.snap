// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <DashboardRoutes /> debe mostrarse el componenete correctamente 1`] = `"<nav class=\\"navbar navbar-expand-sm navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"/\\">Asociaciones</a><div class=\\"navbar-collapse\\"><div class=\\"navbar-nav\\"><a class=\\"nav-item nav-link active\\" href=\\"/marvel\\" aria-current=\\"page\\">Marvel</a><a class=\\"nav-item nav-link\\" href=\\"/dc\\">DC</a><a class=\\"nav-item nav-link\\" href=\\"/search\\">Search</a></div></div><div class=\\"navbar-collapse collapse w-100 order-3 dual-collapse2\\"><ul class=\\"navbar-nav ml-auto\\"><span class=\\"nav-item nav-link text-info\\"> Claudia </span><button class=\\"nav-item nav-link btn\\">Logout</button></ul></div></nav><div class=\\"container mt-3\\"><div><h1>Marvel Screen</h1><hr><div class=\\"card-columns animate__animated animate__pulse\\"><div class=\\"card ms-3\\" style=\\"max-width: 540px;\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4 \\"><img src=\\"./assets/heroes/marvel-spider.jpg\\" alt=\\"Spider Man\\" class=\\"card-img\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><h5 class=\\"card-title\\">Spider Man</h5><p class=\\"card-text\\">Peter Parker</p><p class=\\"card-text\\"><small class=\\"text-muted\\">Amazing Fantasy #15</small></p><a href=\\"/hero/marvel-spider\\">Más ....</a></div></div></div></div><div class=\\"card ms-3\\" style=\\"max-width: 540px;\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4 \\"><img src=\\"./assets/heroes/marvel-captain.jpg\\" alt=\\"Captain America\\" class=\\"card-img\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><h5 class=\\"card-title\\">Captain America</h5><p class=\\"card-text\\">Steve Rogers</p><p class=\\"card-text\\"><small class=\\"text-muted\\">Captain America Comics #1</small></p><a href=\\"/hero/marvel-captain\\">Más ....</a></div></div></div></div><div class=\\"card ms-3\\" style=\\"max-width: 540px;\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4 \\"><img src=\\"./assets/heroes/marvel-iron.jpg\\" alt=\\"Iron Man\\" class=\\"card-img\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><h5 class=\\"card-title\\">Iron Man</h5><p class=\\"card-text\\">Tony Stark</p><p class=\\"card-text\\"><small class=\\"text-muted\\">Tales of Suspense #39</small></p><a href=\\"/hero/marvel-iron\\">Más ....</a></div></div></div></div><div class=\\"card ms-3\\" style=\\"max-width: 540px;\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4 \\"><img src=\\"./assets/heroes/marvel-thor.jpg\\" alt=\\"Thor\\" class=\\"card-img\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><h5 class=\\"card-title\\">Thor</h5><p class=\\"card-text\\">Thor Odinson</p><p class=\\"card-text\\"><small class=\\"text-muted\\">Journey into Myster #83</small></p><a href=\\"/hero/marvel-thor\\">Más ....</a></div></div></div></div><div class=\\"card ms-3\\" style=\\"max-width: 540px;\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4 \\"><img src=\\"./assets/heroes/marvel-hulk.jpg\\" alt=\\"Hulk\\" class=\\"card-img\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><h5 class=\\"card-title\\">Hulk</h5><p class=\\"card-text\\">Bruce Banner</p><p class=\\"card-text\\"><small class=\\"text-muted\\">The Incredible Hulk #1</small></p><a href=\\"/hero/marvel-hulk\\">Más ....</a></div></div></div></div><div class=\\"card ms-3\\" style=\\"max-width: 540px;\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4 \\"><img src=\\"./assets/heroes/marvel-wolverine.jpg\\" alt=\\"Wolverine\\" class=\\"card-img\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><h5 class=\\"card-title\\">Wolverine</h5><p class=\\"card-text\\">James Howlett</p><p class=\\"card-text\\"><small class=\\"text-muted\\">The Incredible Hulk #180</small></p><a href=\\"/hero/marvel-wolverine\\">Más ....</a></div></div></div></div><div class=\\"card ms-3\\" style=\\"max-width: 540px;\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4 \\"><img src=\\"./assets/heroes/marvel-daredevil.jpg\\" alt=\\"Daredevil\\" class=\\"card-img\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><h5 class=\\"card-title\\">Daredevil</h5><p class=\\"card-text\\">Matthew Michael Murdock</p><p class=\\"card-text\\"><small class=\\"text-muted\\">Daredevil #1</small></p><a href=\\"/hero/marvel-daredevil\\">Más ....</a></div></div></div></div><div class=\\"card ms-3\\" style=\\"max-width: 540px;\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4 \\"><img src=\\"./assets/heroes/marvel-hawkeye.jpg\\" alt=\\"Hawkeye\\" class=\\"card-img\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><h5 class=\\"card-title\\">Hawkeye</h5><p class=\\"card-text\\">Clinton Francis Barton</p><p class=\\"card-text\\"><small class=\\"text-muted\\">Tales of Suspense #57</small></p><a href=\\"/hero/marvel-hawkeye\\">Más ....</a></div></div></div></div><div class=\\"card ms-3\\" style=\\"max-width: 540px;\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4 \\"><img src=\\"./assets/heroes/marvel-cyclops.jpg\\" alt=\\"Cyclops\\" class=\\"card-img\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><h5 class=\\"card-title\\">Cyclops</h5><p class=\\"card-text\\">Scott Summers</p><p class=\\"card-text\\"><small class=\\"text-muted\\">X-Men #1</small></p><a href=\\"/hero/marvel-cyclops\\">Más ....</a></div></div></div></div><div class=\\"card ms-3\\" style=\\"max-width: 540px;\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4 \\"><img src=\\"./assets/heroes/marvel-silver.jpg\\" alt=\\"Silver Surfer\\" class=\\"card-img\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><h5 class=\\"card-title\\">Silver Surfer</h5><p class=\\"card-text\\">Norrin Radd</p><p class=\\"card-text\\"><small class=\\"text-muted\\">The Fantastic Four #48</small></p><a href=\\"/hero/marvel-silver\\">Más ....</a></div></div></div></div></div></div></div>"`;
